src/Features/LanguageServer/Protocol/Features/EditAndContinue/EditAndContinueDiagnosticSource.cs

```diff
@@ -2,19 +2,77 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
 using System.Collections.Immutable;
 using System.Diagnostics;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.EditAndContinue;
 
 internal sealed class EditAndContinueDiagnosticSource
 {
-    public static async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(Document designTimeDocument, CancellationToken cancellationToken)
+    private sealed class ClosedDocumentSource(TextDocument document, ImmutableArray<DiagnosticData> diagnostics) : AbstractWorkspaceDocumentDiagnosticSource(document)
```

Comment:
nit: consider placing in separate file.

Reply:
I think that's unnecessary. It's just 7 lines per type.